Fix/nex 91/extension paths resolver

oatymart opened the pull request:
In order to bundle extensions **which depend on** the first vendorised extensions (`taoItems` and `taoQtiItem`) without errors, the new npm paths from these dependency extensions are needed inside the `amdResolve` tool. I am able to load these paths from a `paths.json` file in extensions which provide one.

I also had to change the RegExps which determine if the extension prefix should be added back when resolving a path. Because we now have paths like `../../../taoItems/views/node_modules...`, these were being wrongly prefixed in 2 cases, resulting in something like `taoItems/../../../taoItems/views/node_modules...` or `taoQtiTestPreviewer/../../../taoItems/views/node_modules...` which did not survive later string transformations.

The `paths.json` contents are also now used to exclude dependency-owned modules from the bundles: bundle sizes are consistent with what I generated before these changes.

To test:
- clone this branch and `npm link` it in place of the `grunt-tao-bundle` which was installed in TAO from `/tao/views/build`.
- checkout these branches:
  - https://github.com/oat-sa/extension-tao-item/pull/371
  - https://github.com/oat-sa/extension-tao-itemqti/pull/1299
- bundle some extensions with `npx grunt tao[extension]bundle`, and use TAO in non-debug mode (the more complex bundles are `tao`, `taoQtiTest`, `taoQtiTestPreviewer`, `taoClientDiagnostic`)
- also, `npm test` to run unit tests in this repo's folder

oatymart committed:
Moved pathConfig to upper scope so extensions bundlers can define paths

oatymart committed:
More flexible approach to extension-in-path regexes

oatymart committed:
New test assets

oatymart committed:
New test assets - moved

oatymart committed:
Change dependency path loading

oatymart committed:
Inject dependencies into amdResolve

oatymart committed:
Construct dependency regexps + refactor patterns + needsPrefix function

oatymart committed:
Added 2 more bundler tests

oatymart committed:
Added rootPath option to allow to find files easier in tests vs real-world

oatymart committed:
Try to exclude modules based on dependencies paths.json

oatymart committed:
Version bump

btamas commented on test/bundler_spec.js:
```diff
@@ -39,7 +39,8 @@ describe('bundler', () => {
         rootExtension: 'extA',
-        workdir : 'output',
+        workdir : path.join(dataDir, 'output'),
+        rootPath: path.resolve(__dirname, 'data'),
```
This is `dataDir`.

oatymart replied:
ok

btamas commented on lib/amdResolve.js:
```diff
@@ -89,8 +91,16 @@ module.exports = function resolve(pattern, {
-    const extensionPattern = new RegExp(`^${targetExtension}/`);
+    //define regular expressions to use on file path
+    const patterns = {
+        endsWithFileType: new RegExp(`\\.${fileExtension}$`),
```
I'm not a fan of using regular expression for these simple cases.

I think this one is faster and easier.
`dependency.endsWith(fileExtension)`

btamas commented on lib/amdResolve.js:
```diff
@@ -120,6 +130,23 @@ module.exports = function resolve(pattern, {
+                if(!extensionPrefix) {
+                    return false;
+                }
+                if (patterns.startsWithExtension.test(file) || patterns.containsExtension.test(file)) {
```
`file.startsWith(targetExtension)`

btamas commented on lib/amdResolve.js:
```diff
@@ -120,6 +130,23 @@ module.exports = function resolve(pattern, {
+                    return false;
+                }
+                if (Array.isArray(dependencies) && dependencies.length) {
+                    if (patterns.startsWithDependency.test(file) || patterns.containsDependency.test(file)) {
```
`dependencies.some(dependency => file.startWith(dependency))`

or you can replace both startWitch

`dependencies.some(dependency => file.includes(dependency))`

oatymart replied:
I can't use it here, I need the RegExp functionality of `dependencies.join('|')`

oatymart replied:
ok, I used it!

btamas commented on lib/amdResolve.js:
```diff
@@ -120,6 +130,23 @@ module.exports = function resolve(pattern, {
+                    }
+                }
+                return true;
+            }
```
missing semicolon

btamas commented on lib/amdResolve.js:
```diff
@@ -129,23 +156,23 @@ module.exports = function resolve(pattern, {
-                        if(extensionPrefix && !extensionPattern.test(file)){
-                            moduleName +=  targetExtension + '/';
+                        if(needsPrefix(file)) {
+                            moduleName += targetExtension + '/';
```
String concat.

btamas commented on lib/bundler.js:
```diff
@@ -169,12 +177,27 @@ module.exports = async function bundler({
+        // exclude dependencies from bundles:
+        if (Array.isArray(dependencies)) {
+            //dependency using extension bundles
+            dependencies.forEach( async (dependency) => {
```
Be careful, because you won't wait for the resolve of these promises, so the execution will go forward before all `excludeModules` will be finished. You should use `Promise.all` as we discussed.

oatymart replied:
`await Promise.all()` solution works :)

btamas commented on lib/bundler.js:
```diff
@@ -232,6 +250,13 @@ module.exports = async function bundler({
         if(Array.isArray(dependencies)){
             dependencies.forEach( async (dependency) => {
```
This foreach has the same problem like the other above.

oatymart committed:
Code cleanup

oatymart committed:
Wrapped dependencies loops in Promise.all to achieve correct async flow

oatymart committed:
Replaced RegExps with string methods

ferenckiss89 reviewed (approved):
Good for me.

krampstudio commented on lib/bundler.js:
```diff
@@ -45,15 +45,21 @@ module.exports = async function bundler({
     paths = {},
     workDir = 'output',
     outputDir = 'loader',
+    rootPath = path.resolve('../../..'), // goes up from /tao/views/build
```
I'd prefer not have `path.resolve('../../..')` as a default but in the implementation (`tao/views/build/grunt/bundle.js`) because if it already contains the definition of the paths that will be used in the task

krampstudio commented on lib/bundler.js:
```diff
@@ -159,6 +166,19 @@ module.exports = async function bundler({
+        if (Array.isArray(dependencies)) {
+            await Promise.all(dependencies.map(dependency => {
+                try {
+                    const dependencyPaths = require(`${rootPath}/${dependency}/views/build/grunt/paths.json`);
```
since it used at multiple places, would you mind create a function `loadExtensionPaths` or `loadExtensionExtraPaths` , `loadExtensionAliases`, ...

krampstudio commented on package.json:
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@oat-sa/grunt-tao-bundle",
-  "version": "0.2.0",
+  "version": "0.2.1",
```
the API has changed it's a feature version `0.3.0`

krampstudio reviewed (changes requested):
Look very good. 
I've just a few possible improvements and I've still a doubt on the generated modules.
It looks like bundles contains the dependencies from tao-core-sdk and tao-core-ui when they should be excluded (not sure if it comes from your changes or if the issue was already there). Maybe tao-core should also have a `paths.json` so they would be excluded. But here we are loading the some module multiple times ( tao-core-sdk and tao-core-ui  are part of vendor and always excluded)

oatymart committed:
Extracted paths.json require to a reusable function

oatymart committed:
Set rootPath only in consumer

oatymart committed:
Log what goes into each bundle

oatymart committed:
Make vendor modules not take prefix, so they are excluded correctly

oatymart committed:
Version bump 0.3.0

oatymart commented:
By changing the paths in `requirejs.build.json` (here https://github.com/oat-sa/tao-core/pull/2209), I've fixed the issue where the resolver was unable to resolve the modules in `/tao/views/node_modules`.

I also added a 1-line check which prevents `amd.vendor` defined modules from being prefixed and mis-resolved, and now they are properly excluded from the non-vendor bundles.

Some external CSS still makes it into the bundles, I didn't solve that yet.

Edit: there is also a new logging output during the bundles creation. I think it's helpful during this refactoring phase, but I can remove it once we are happy with the bundling process again.

btamas commented on lib/bundler.js:
```diff
@@ -169,13 +180,12 @@ module.exports = async function bundler({
-                try {
-                    const dependencyPaths = require(`${rootPath}/${dependency}/views/build/grunt/paths.json`);
+                const dependencyPaths = loadExtensionPaths(dependency);
+                if (Object.keys(dependencyPaths).length) {
```
`Object.keys(dependencyPaths)` should be a variable, because you use it multiple times.

btamas commented on lib/bundler.js:
```diff
@@ -245,14 +255,12 @@ module.exports = async function bundler({
-                        const dependencyPaths = require(`${rootPath}/${dependency}/views/build/grunt/paths.json`);
+                    // load the requirejs paths of each dependency into the global pathConfig
+                    const dependencyPaths = loadExtensionPaths(dependency);
+                    if (Object.keys(dependencyPaths).length) {
```
Useless condition. Everything works without it.

btamas reviewed (commented):
Just small improvements.

ferenckiss89 commented on lib/amdResolve.js:
```diff
@@ -120,6 +119,24 @@ module.exports = function resolve(pattern, {
+                if (file.startsWith(targetExtension) || file.includes(`/${targetExtension}/`)) {
+                    return false;
+                }
+                if (Array.isArray(dependencies) && dependencies.length) {
```
these two conditions can be combined since there is no else condition

oatymart replied:
I did it that way for clearer separation of the different logical tests. The more conditions I put into one `if()`, the longer it takes to understand.

ferenckiss89 replied:
Yes but we should consider this always as a code smell (imo) or as unnecessary nesting, there is always a way to reduce the amount of the conditions or to invert the logic:

if (!Array.isArray(dependencies) || !dependencies.length) {
    return true;
}
if (dependencies.some(dep => file.startsWith(dep)) ||
     dependencies.some(dep => file.includes(`/${dep}/`))) {
     return false;
}

ferenckiss89 reviewed (commented):
Minor comment.

oatymart committed:
Simplify dependencyPaths usages

oatymart committed:
Take out console.logs

oatymart committed:
Exclude vendor css by adding amdLoader later

oatymart committed:
Refactor addModules to ease debugging

oatymart committed:
Fix logical test on forbidden module check

oatymart committed:
Tests & assets for throwing on forbidden module check

btamas commented on test/data/extB/views/node_modules/@oat-sa/some-fake-library/dist/another-fake-file.js:
```diff
@@ -0,0 +1,7 @@
+define([
+    'css!extB/some-fake-lib/css/some-fake-style',
```
Maybe you can extend the test with a dependency that contains `.css` and `.tpl` in the end to be sure it will work as well.

ferenckiss89 replied:
For now, this covers the main functionality. Can be added later.

ferenckiss89 commented on lib/bundler.js:
```diff
@@ -81,21 +81,37 @@ module.exports = async function bundler({
+     * @returns {Array}
+     */
+    const resolveModules = async (modulePatterns, fileType) => {
+        var allResolved = [];
```
const

ferenckiss89 replied:
will be changed later
